mlir/lib/Dialect/Rock/Transforms/ReuseLDS.cpp

```diff
@@ -0,0 +1,441 @@
+    requiredMemory += size;
+  }
+
+  // not enough LDS memory
+  if (failed(checkLDSSize(func, requiredMemory))) {
+    LLVM_DEBUG(llvm::dbgs() << "ReuseLDS requires too much LDS memory: "
+                            << requiredMemory << " bytes\n");
+    return failure();
+  }
+
+  // write the new GpuAllocOp to the start
+  rewriter.setInsertionPointToStart(&func.getBody().front());
+
+  // New allocations
+  llvm::MapVector<int64_t, GpuAllocOp> colorAllocs;
+  auto workgroupMemoryAddressSpace = rewriter.getAttr<gpu::AddressSpaceAttr>(
+      gpu::GPUDialect::getWorkgroupAddressSpace());
+  for (auto [color, size] : colorSizes) {
+    auto ldsMemRefBufferType = MemRefType::get(
+        {size}, rewriter.getI8Type(), AffineMap{}, workgroupMemoryAddressSpace);
+
+    Location loc = func.getLoc();
+    GpuAllocOp colorAlloc =
+        rewriter.create<GpuAllocOp>(loc, ldsMemRefBufferType);
+    colorAllocs[color] = colorAlloc;
+    LLVM_DEBUG(llvm::dbgs()
+               << "allocating " << size << " bytes, color: " << color << "\n");
+  }
+
+  // Replace all GpuAllocs as ViewOps
+  for (auto allocTuple : allocOffsets) {
+    GpuAllocOp alloc;
+    int64_t color, offset;
+    bool useLDSBarrier;
+    std::tie(alloc, color, offset, useLDSBarrier) = allocTuple;
+    int64_t colorSize = colorSizes[color];
+    GpuAllocOp newAlloc = colorAllocs[color];
+
+    if (offset < 0) {
+      LLVM_DEBUG(llvm::dbgs() << "Negative offset\n");
+      return failure();
+    }
+    if (offset >= colorSize) {
+      LLVM_DEBUG(llvm::dbgs() << "Offset is too big\n");
+      return failure();
+    }
+    auto bufferType = alloc.getOutput().getType();
+    auto numElements = bufferType.getNumElements();
+    auto elementBytes = getByteWidth(bufferType.getElementType());
+    auto rank = bufferType.getRank();
+    if (elementBytes != 1) {
+      LLVM_DEBUG(llvm::dbgs() << "GpuAllocOp allocates a type of "
+                              << elementBytes << " bytes, expected 1 byte\n");
+      return failure();
+    }
+    if (rank != 1) {
+      LLVM_DEBUG(llvm::dbgs() << "Rank should be one, it's " << rank << "\n");
+      return failure();
+    }
+
+    rewriter.setInsertionPointAfter(alloc);
+    Location loc = alloc->getLoc();
+
+    LLVM_DEBUG(llvm::dbgs() << "color: " << color << " offset: " << offset
+                            << " numElements: " << numElements << "\n");
+    Value byteOffset =
+        rewriter.createOrFold<arith::ConstantIndexOp>(loc, offset);
+
+    auto newViewType =
+        MemRefType::get({numElements}, rewriter.getI8Type(), AffineMap{},
+                        workgroupMemoryAddressSpace);
+    rewriter.replaceOpWithNewOp<memref::ViewOp>(alloc, newViewType, newAlloc,
+                                                byteOffset, ValueRange{});
+
+    // add barrier if needed
+    if (useLDSBarrier) {
+      rewriter.create<LDSBarrierOp>(loc);
+    }
+  }
+
+  // Remove all GpuDeallocOps but the last one
```

Comment:
`// Remove all GPuDeallocOps but the last one and add a new alloc/dealloc pair for each buffer`